File: process/R/data/output.R

create_outputs <- function(data_dir) {
  data_path <- file.path(data_dir, "results.parquet")
  data <- read_parquet(data_path) %>%
    as_tibble()
  
  # <><><><><><><><><><><><><><><><><><><><>
  # 1. Produce population status
  # <><><><><><><><><><><><><><><><><><><><>
  pop_stats <- data %>%
    group_by(year, life_stage) %>%
    summarise(count = n(), .groups = "drop")
  
  hhd_stats <- data %>%
    # Filter for alive people (data[data["life_stage"] == "alive"])
    filter(life_stage == "alive") %>%
    # Select year and household_id and get unique combinations (.drop_duplicates())
    distinct(year, household_id) %>%
    # Group by year
    group_by(year) %>%
    # Count the number of unique household IDs (groupby("year")["household_id"].count())
    summarise(count = n(), .groups = "drop")
  
  # The return structure in R is typically a named list
  return(list(
    population = list(
      pop = pop_stats,
      hhd = hhd_stats
    )
  ))
}

File: process/Python/data/output.py
from os.path import join
from pyarrow.parquet import read_table as pq_read_table


def create_outputs(data_dir: str):
    data_path = join(data_dir, "results.parquet")
    data = pq_read_table(data_path)
    data = data.to_pandas()

    # <><><><><><><><><><><><><><><><><><><><>
    # 1. Produce population status
    # <><><><><><><><><><><><><><><><><><><><>
    pop_stats = data.groupby(["year", "life_stage"]).size().reset_index(name="count")
    hhd_stats = (
        data[data["life_stage"] == "alive"][["year", "household_id"]]
        .drop_duplicates()
        .groupby("year")["household_id"]
        .count()
        .reset_index(name="count")
    )

    # <><><><><><><><><><><><><><><><><><><><>
    # 2. Produce employment status
    # <><><><><><><><><><><><><><><><><><><><>
    employment_stats = {}
    for proc_key in ["market_income", "latent_market_income"]:
        employment_stats[f"mean_{proc_key}"] = (
            data[(data["life_stage"] == "alive")]
            .groupby("year")[proc_key]
            .mean()
            .reset_index(name=f"mean_{proc_key}")
        )
    return {
        "population": {"pop": pop_stats, "hhd": hhd_stats},
        "employment": {"income": employment_stats},
    }
